backend/app/services/pipeline_service.py

```diff
@@ -90,3 +94,59 @@ def delete_pipeline_by_id(self, pipeline_id: UUID) -> None:
                 raise ResourceInUseError(ResourceType.PIPELINE, str(pipeline_id), MSG_ERR_DELETE_RUNNING_PIPELINE)
             self._persistence.delete_by_id(pipeline_id, db)
             db.commit()
+
+    def get_pipeline_metrics(self, pipeline_id: UUID, duration_seconds: int = 60) -> PipelineMetrics:
+        """Calculate metrics for a pipeline over a specified time window."""
+        # First check if pipeline exists
+        pipeline = self.get_pipeline_by_id(pipeline_id)
+        if not pipeline.model_id:
+            raise ValueError("Cannot get metrics for a pipeline without an associated model.")
+
+        # Calculate time window
+        end_time = datetime.now(UTC)
+        start_time = end_time - timedelta(seconds=duration_seconds)
+
+        # Get actual latency measurements from the metrics collector
+        metrics_collector = MetricsCollector()
```

Comment:
Can you move the metrics collector initialization to the constructor? Then you don't have to make the class Singleton, I guess.

Reply:
The original comment is still relevant. There are two possible approaches to avoid instantiating extra objects within this method (both options help prevent monkey-patching in tests and allow for better composition of fixtures):

- **Option 1:** Create a shared memory reference in the constructor and access it directly to retrieve metrics.
- **Option 2:** Pass a `MetricsCollector` instance via dependencies (see `dependencies.py`), ensuring a single instance exists throughout the web server lifecycle.

I believe the retrieval logic should be decoupled from `MetricsCollector`, as its name suggests a focus on collection. Retrieval is only required within `PipelineService`.